esrally/mechanic/supplier.py

```diff
@@ -328,11 +328,19 @@ def cached(self):
         return self.cached_path is not None and os.path.exists(self.cached_path)
 
     def fetch(self):
-        resolved_revision = self.source_supplier.fetch()
-        if resolved_revision:
-            # ensure we use the resolved revision for rendering the artifact
-            self.file_resolver.revision = resolved_revision
-            self.cached_path = os.path.join(self.distributions_root, self.file_name)
+        # Can we already resolve the artifact without fetching the source tree at all? This is the case when a specific
+        # revision (instead of a meta-revision like "current") is provided and the artifact is already cached. This is
+        # also needed if an external process pushes artifacts to Rally's cache which might have been built from a
+        # fork. In that case the provided commit hash would not be present in any case in the main ES repo.
+        maybe_an_artifact = os.path.join(self.distributions_root, self.file_name)
+        if os.path.exists(maybe_an_artifact):
+            self.cached_path = maybe_an_artifact
+        else:
+            resolved_revision = self.source_supplier.fetch()
+            if resolved_revision:
+                # ensure we use the resolved revision for rendering the artifact
+                self.file_resolver.revision = resolved_revision
+                self.cached_path = os.path.join(self.distributions_root, self.file_name)
```

Comment:
Ha, good catch but that's tricky. The point is that `self.file_name` is not a constant value but a (computed) property that relies on the `file_resolver`. In the first call (where we assign it to `maybe_an_artifact`) we just use the user-provided revision (if any) that is provided with `--revision`. This could be a valid git revision hash (like `d31808de80f`) but it could also be a meta-revision (like `current`). So in the first case it might resolve the artifact file name to `elasticsearch-d31808de80f-linux-x86_64.tar.gz`) and then we could indeed assign `maybe_an_artifact` here. But in the second case it would resolve it to `elasticsearch-current-linux-x86_64.tar.gz`) and that's not what we want. Instead, we need to find out what `current` means and resolve it to a real git revision hash that we can use for the file name of the cached artifact. We do this here:

```python
resolved_revision = self.source_supplier.fetch()
if resolved_revision:
    # ensure we use the resolved revision for rendering the artifact
    self.file_resolver.revision = resolved_revision
```

The `source_supplier` fetches the source tree (if needed) based on the provided revision (may also be `current`, `latest` or a timestamp) and returns the *real*, resolved git revision. We then provide that git revision to our file resolver which in turn will use it to return the corresponding file name.